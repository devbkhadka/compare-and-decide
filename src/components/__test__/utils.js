import React from 'react'
import { Provider } from 'react-redux'
import { IntlProvider } from 'react-intl'
import { render } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'

const middlewares = []
const mockStore = configureMockStore(middlewares)

export function renderWithIntl(component) {
    return render(<IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>)
}

export function renderWithProvider(component, initialState={}) {
    const store = mockStore(initialState)
    return [render(<Provider store={store}>
        <IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>
    </Provider>), store]
}