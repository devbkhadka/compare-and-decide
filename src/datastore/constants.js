export const ADD_ITEM_WITH_TITLE = 'ADD_ITEM_WITH_TITLE'