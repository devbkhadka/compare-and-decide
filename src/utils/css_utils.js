export function joinClss(...classes){
    return classes.join(' ')
}