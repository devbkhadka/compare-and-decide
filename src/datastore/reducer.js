import { ADD_ITEM_WITH_TITLE } from './constants'

const initialState = {
    items:[]
}

export default function (state=initialState, payload) {
    switch(payload.type) {
        case ADD_ITEM_WITH_TITLE:
            if(payload.data){
                return {...state, items: [...state.items, payload.data]}
            }
        
        default:
            return state
    }
}