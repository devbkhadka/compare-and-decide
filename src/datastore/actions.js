import { UPDATE_ITEMS, 
    UPDATE_LANGUAGE, 
    UPDATE_ATTRIBUTES,
    UPDATE_PREFERRED_ITEM
} from './constants'

import {COMPARISION_STATUS} from './reducer'

export const ITEMS_KEY = 'items-key'
export const ATTRIBUTES_KEY = 'attributes-key'
export const PREFETED_ITEM_TITLE_KEY = 'prefered-item-key'

export function addItemWithTitle(title) {
    return (dispatch, getState) => {

        const state = getState()
        if(!title || state.items[title]) {
            return
        }

        const item = {
            title: title,
            values: {}
        }
        state.items[item.title] = item
        updateAndSaveItems(state.items, dispatch)
    }
}

export function loadStateFromStorage() {
    return dispatch => {
        const itemsJson = localStorage.getItem(ITEMS_KEY)
        if(itemsJson) {
            dispatch(updateItems(JSON.parse(itemsJson)))
        }

        const attributesJson = localStorage.getItem(ATTRIBUTES_KEY)
        if(attributesJson) {
            dispatch(updateAttributes(JSON.parse(attributesJson)))
        }

        const preferedItemTitle = localStorage.getItem(PREFETED_ITEM_TITLE_KEY)
        if(preferedItemTitle){
            dispatch(updatePreferedItemTitle(preferedItemTitle))
        }
    }
}

export function updateItems(items) {
    return {
        type: UPDATE_ITEMS,
        data: items
    }
}

export function updateLanguage(language) {
    return {
        type: UPDATE_LANGUAGE,
        data: language
    }
}

export function addAttribute(attribute) {
    return (dispatch, getState)=>{
        const attributes = getState().attributes
        if(attribute && attributes.indexOf(attribute)===-1) {
            attributes.push(attribute)

            updateAndSaveAttributes(attributes, dispatch)
        }
    }
}

export function updateAttributes(attributes) {
    return {
        type: UPDATE_ATTRIBUTES,
        data: attributes
    }
}

export function updateItemAttributeValue(item, attr, value) {
    return (dispatch, getState) => {
        const state = getState()
        const newItems = {...state.items}
        newItems[item.title] = {...item, values: {...item.values, [attr]: value}}
        
        updateAndSaveItems(newItems, dispatch)
    }
}

export function setPreferedItemTitle(itemTitle) {
    return (dispatch, getState) => {
        const state = getState()
        const curPreferedTitle = state.preferedItemTitle

        const item = state.items[itemTitle]
        if(!item) {
            // TODO: This branch not tested
            return
        }
        if(curPreferedTitle && state.items[curPreferedTitle]){
            // TODO: This branch not tested
            state.items[curPreferedTitle].comparisionStatus = undefined
        }
        item.comparisionStatus = COMPARISION_STATUS.ITEM_PREFERED

        updateAndSaveItems(state.items, dispatch)
        updateAndSavePreferedItemTitle(itemTitle, dispatch)
        
    }
}

export function updatePreferedItemTitle(itemTitle) {
    return {
        type: UPDATE_PREFERRED_ITEM,
        data: itemTitle
    }
}

export function rejectItem(itemTitle){
    return (dispatch, getState) => {
        const state = getState()
        state.items[itemTitle].comparisionStatus = COMPARISION_STATUS.ITEM_REJECTED

        updateAndSaveItems(state.items, dispatch)
    }
}

export function deleteAttribute(attribute) {
    return (dispatch, getState) => {
        const {items, attributes} = getState()
        const newAttributes = attributes.filter(attr=>attribute !== attr)

        for(const key in items) {
            const values = items[key].values
            delete values[attribute]
        }

        updateAndSaveItems(items, dispatch)
        updateAndSaveAttributes(newAttributes, dispatch)
    }
}

export function deleteItem(itemTitle) {
    return (dispatch, getState) => {
        const {items} = getState()
        delete items[itemTitle]

        updateAndSaveItems(items, dispatch)
    }
}

export function clearComparisionStatus(itemTitle) {
    return (dispatch, getState) => {
        const {items, preferedItemTitle} = getState()
        items[itemTitle].comparisionStatus = undefined
        updateAndSaveItems(items, dispatch)
        if(preferedItemTitle===itemTitle) {
            updateAndSavePreferedItemTitle(undefined, dispatch)
        }
    }
}


function updateAndSaveItems(items, dispatch) {
    localStorage.setItem(ITEMS_KEY, JSON.stringify(items))
    dispatch(updateItems(items))
}

function updateAndSavePreferedItemTitle(itemTitle, dispatch) {
    localStorage.setItem(PREFETED_ITEM_TITLE_KEY, itemTitle)
    dispatch(updatePreferedItemTitle(itemTitle))
}

function updateAndSaveAttributes(attributes, dispatch) {
    localStorage.setItem(ATTRIBUTES_KEY, JSON.stringify(attributes))
    dispatch(updateAttributes(attributes))
}