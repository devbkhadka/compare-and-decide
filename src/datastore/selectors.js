import { COMPARISION_STATUS } from './reducer'

const itemsSelector = (items, selector)=>{
    const selectedItems = []
    for(const key of Object.keys(items)){
        const item = items[key]
        if(selector(item)){
            selectedItems.push(item)
        }
    }
    return selectedItems
}

export const remainingItemsSelector = state=> itemsSelector(state.items, item=>!item.comparisionStatus)
export const rejectedItemsSelector = state =>
    itemsSelector(state.items, item=> item.comparisionStatus===COMPARISION_STATUS.ITEM_REJECTED)