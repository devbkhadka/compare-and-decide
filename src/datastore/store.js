import {createStore, applyMiddleware, compose} from 'redux'
import ReduxThunk from 'redux-thunk'
import reducer from './reducer'


export const middleWares = [ReduxThunk]

const devToolEnhancer = (window && window.__REDUX_DEVTOOLS_EXTENSION__ && 
                            window.__REDUX_DEVTOOLS_EXTENSION__())
let enhancer = applyMiddleware(...middleWares)
if(devToolEnhancer) {
    enhancer = compose(enhancer, devToolEnhancer)
} 

const store = createStore(reducer, enhancer)

export default store
