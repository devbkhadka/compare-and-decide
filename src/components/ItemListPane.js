import React, {useState, useEffect} from 'react'
import { useSelector, useDispatch } from 'react-redux'
import {
    TextField, 
    Tabs, 
    Tab, 
    Box,
    IconButton,
    Typography,
    withStyles,
    useTheme,
} from '@material-ui/core'
import {AddCircleRounded as Add} from '@material-ui/icons'
import { addItemWithTitle } from '../datastore/actions'

const StyledTab = withStyles(theme=>({
    wrapper: {
        flexDirection: 'row',
        justifyContent: 'flex-start'
    },
    wrapped: {
        borderBottom: `1px solid ${theme.palette.divider}`,
        '&.Mui-selected': {
            backgroundColor: theme.palette.secondary.main,
            color: theme.palette.secondary.contrastText
        }
    }
})) (Tab)

export default function ItemListPane({ onItemSelected }) {
    const [itemTitle, setItemTitle] = useState('')
    const items = useSelector(state=>state.items)
    const dispatch = useDispatch()

    const [activeTab, setActiveTab] = useState(0)
    const handleItemTitleChanged = (e)=>{
        setItemTitle(e.target.value)
    }
    const handleAdd = (e) => {
        dispatch(addItemWithTitle(itemTitle))
        setItemTitle('')
    }

    const handleActiveItemChanged = (e, newValue) => {
        setActiveTab(newValue)
        onItemSelected && onItemSelected(items[newValue])
    }

    useEffect(()=>{
        if(items.length===1){
            onItemSelected && onItemSelected(items[0])
        }
    })
    const theme = useTheme()
    return <>
        <Typography variant='h5' style={{marginBottom: theme.spacing(1)}}>
            Items
        </Typography>
        <Box display='flex' style={{marginRight: theme.spacing(1)}}>
            <TextField style={{flexGrow:1}} required data-testid='inpNewitem' 
                value={itemTitle} onChange={ handleItemTitleChanged }
                placeholder='Add New Item'/>
            <IconButton data-testid='btnAdd' onClick={handleAdd}>
                <Add>Add</Add>
            </IconButton>
        </Box>
        <Tabs orientation="vertical" variant="scrollable" value={activeTab}
            onChange={handleActiveItemChanged}>
            {
                items.map((item, i)=><StyledTab key={i} label={item} wrapped/>)
            }
        </Tabs>
    </>
}