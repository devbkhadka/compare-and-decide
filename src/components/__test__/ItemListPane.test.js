import React from 'react'
import { renderWithProvider } from './utils'
import { fireEvent } from '@testing-library/react'
import { addItemWithTitle } from '../../datastore/actions'
import ItemListPane from '../ItemListPane'

const render = (component, state={items:[]})=> renderWithProvider(component, state)

describe("ItemListPane Test", ()=>{
    it('should dispatch addItemWithTitle action when add button clicked', async ()=>{
        const [rendered, store] = render(<ItemListPane></ItemListPane>)

        const input = rendered.getByPlaceholderText('Add New Item')
        const addButton = rendered.getByTestId('btnAdd')
        
        fireEvent.change(input, {target:{value: 'Item 1'}})
        fireEvent.click(addButton)

        fireEvent.change(input,  {target:{value: 'Item 2'}})
        fireEvent.click(addButton)

        expect(store.getActions()).toEqual([addItemWithTitle('Item 1'), addItemWithTitle('Item 2')])
    })

    it('should match saved snapshot', ()=>{
        const [rendered] = render(<ItemListPane></ItemListPane>)
        expect(rendered.baseElement).toMatchSnapshot()
    })

    it('should render items from redux store', ()=>{
        const items = {
            ['item1']: { 
                title: 'Item1',
            }, 
            ['item2']: {
                title: 'Item2',
            }
        }
        const [rendered] = render(<ItemListPane></ItemListPane>, {items})

        rendered.getByText('Item1')
        rendered.getByText('Item2')
    })

    it('should trigger onItemSelected event when a item is clicked', ()=>{
        const items = {
            ['item1']: { 
                title: 'Item1',
            }, 
            ['item2']: {
                title: 'Item2',
            }
        }
        const mockOnItemSelected = jest.fn()
        const [rendered] = render(<ItemListPane onItemSelected={mockOnItemSelected}/>, {items})
        fireEvent.click(rendered.getByText('Item2'))

        expect(mockOnItemSelected).toBeCalledWith(items['item2'])
    })

    it('should trigger onItemSelected for 1st item when only one item exists', async ()=>{
        const mockOnItemSelected = jest.fn()
        const items = {
            ['item1']: { 
                title: 'Item1',
            }
        }
        const [rendered] = render(<ItemListPane onItemSelected={mockOnItemSelected}/>, {items})
        
        expect(mockOnItemSelected).toBeCalledWith({title: 'Item1'})
    })
})