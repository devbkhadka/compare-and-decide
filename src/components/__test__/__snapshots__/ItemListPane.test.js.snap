// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemListPane Test should match saved snapshot 1`] = `
<body>
  <div>
    <h5
      class="MuiTypography-root MuiTypography-h5"
      style="margin-bottom: 8px;"
    >
      Items
    </h5>
    <div
      class="MuiBox-root MuiBox-root-147"
      style="margin-right: 8px;"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
        data-testid="inpNewitem"
        style="flex-grow: 1;"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            placeholder="Add New Item"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        data-testid="btnAdd"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm4 11h-3v3c0 .55-.45 1-1 1s-1-.45-1-1v-3H8c-.55 0-1-.45-1-1s.45-1 1-1h3V8c0-.55.45-1 1-1s1 .45 1 1v3h3c.55 0 1 .45 1 1s-.45 1-1 1z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiTabs-root MuiTabs-vertical"
    >
      <div
        class="MuiTabs-scrollable"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-scrollable"
        style="margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer MuiTabs-flexContainerVertical"
          role="tablist"
        />
        <span
          class="PrivateTabIndicator-root-227 PrivateTabIndicator-colorSecondary-229 MuiTabs-indicator PrivateTabIndicator-vertical-230"
          style="top: 0px; height: 0px;"
        />
      </div>
    </div>
  </div>
</body>
`;
